worker/changestream/eventmultiplexer/eventmultiplexer.go

```diff
@@ -0,0 +1,349 @@
+	// run one at a time asynchronously. Allowing changes within a given
+	// term to be signaled of a change independently from one another.
+	// Once a change within a term has been completed, only at that point
+	// is another change processed, until all changes are exhausted.
+	Terms() <-chan changestream.Term
+}
+
+type eventFilter struct {
+	subscriptionID uint64
+	changeMask     changestream.ChangeType
+	filter         func(changestream.ChangeEvent) bool
+}
+
+type reportRequest struct {
+	data map[string]any
+	done chan struct{}
+}
+
+// EventMultiplexer defines an event listener and dispatcher for db changes that
+// can be multiplexed to subscriptions. The event queue allows consumers to
+// subscribe via callbacks to the event queue. This is a lockless
+// implementation, all subscriptions and changes are serialized in the main
+// loop. Dispatching is randomized to ensure that subscriptions don't depend on
+// ordering. The subscriptions can be associated with different subscription
+// options, which provide filtering when dispatching. Unsubscribing is provided
+// per subscription, which is done asynchronously.
+type EventMultiplexer struct {
+	catacomb catacomb.Catacomb
+	stream   Stream
+	logger   Logger
+	clock    clock.Clock
+
+	subscriptions      map[uint64]*subscription
+	subscriptionsByNS  map[string][]*eventFilter
+	subscriptionsAll   map[uint64]struct{}
+	subscriptionsCount uint64
+	dispatchErrorCount int
+
+	// (un)subscription related channels to serialize adding and removing
+	// subscriptions. This allows the queue to be lock less.
+	subscriptionCh   chan subscriptionOpts
+	unsubscriptionCh chan uint64
+
+	reportsCh chan reportRequest
+}
+
+// New creates a new EventMultiplexer that will use the Stream for events.
+func New(stream Stream, clock clock.Clock, logger Logger) (*EventMultiplexer, error) {
+	queue := &EventMultiplexer{
+		stream:             stream,
+		logger:             logger,
+		clock:              clock,
+		subscriptions:      make(map[uint64]*subscription),
+		subscriptionsByNS:  make(map[string][]*eventFilter),
+		subscriptionsAll:   make(map[uint64]struct{}),
+		subscriptionsCount: 0,
+		dispatchErrorCount: 0,
+
+		subscriptionCh:   make(chan subscriptionOpts),
+		unsubscriptionCh: make(chan uint64),
+
+		reportsCh: make(chan reportRequest),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &queue.catacomb,
+		Work: queue.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return queue, nil
+}
+
+// Subscribe creates a new subscription to the event queue. Options can be
+// provided to allow filter during the dispatching phase.
+func (e *EventMultiplexer) Subscribe(opts ...changestream.SubscriptionOption) (changestream.Subscription, error) {
+	// Get a new subscription count without using any mutexes.
+	subID := atomic.AddUint64(&e.subscriptionsCount, 1)
+
+	sub := newSubscription(subID, func() { e.unsubscribe(subID) })
```

Comment:
As I understand it, all the individual dispatches that time out will have their Goroutines running until the multiplexer loop processes their unsubscribe requests.

I suspect it might be lighter on resources to flip a bool in the subscription and have them bumped during gather/dispatch.

What happens in for a subscriber-initiated unsub when events are dispatched after the call, but before the loop processes them? I guess they time out and there's another unsub request passed into the loop...